### PR TITLE
AVRO-2328: Support distinguishing between LocalDateTime and Instant semantics in timestamps

### PR DESCRIPTION
**Pull request by nandorKollar**

### CODE REVIEW

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectLogicalTypes.java [hunk 1]
```diff
@@ -598,6 +601,18 @@ public void testWriteUUIDList() throws IOException {
         read(REFLECT.createDatumReader(stringArraySchema), test).get(0));
   }
 
+  @Test
+  public void testReflectedSchemaLocalDateTime() {
+    Schema actual = REFLECT.getSchema(RecordWithTimestamps.class);
+
+    Assert.assertEquals("Should have the correct record name", "org.apache.avro.reflect", actual.getNamespace());
```
Comment (zivanfi) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectLogicalTypes.java]: It's the namespace, not the name.

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj [hunk 2]
```diff
@@ -242,6 +242,7 @@ TOKEN :
 | < TIME: "time_ms" >
 | < TIMESTAMP: "timestamp_ms" >
 | < DECIMAL: "decimal" >
+| < LOCAL_TIMESTAMP_MS: "local_timestamp_ms" >
```
Comment (zivanfi) [inline lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj]: The other tokens above do not have _MS in the name even if there is a _ms in the string.

Another question: Why are there only tokens for millisec variants but not for microsec variants? This is not only true for the new type but for the old type as well.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericLogicalTypes.java [hunk 3]
```diff
@@ -47,6 +51,8 @@
   public static void addDecimalAndUUID() {
```
Comment (gszadovszky) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericLogicalTypes.java]: After adding the time conversions to this method the name (or the behavior) is really misleading.

File: share/VERSION.txt [hunk 4]
```diff
@@ -1 +1 @@
-1.9.0-SNAPSHOT
```
Comment (gszadovszky) [inline share/VERSION.txt]: I think, this is not related to this change. Shouldn't it be updated automatically after a release? We might create a JIRA for it...

Comment (nandorKollar) [reply share/VERSION.txt]: Indeed it is not related to this issue, however without this change Python tests failed, that's why I added it to the PR.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericLogicalTypes.java (hunk 3 above)
Comment (nandorKollar) [reply lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericLogicalTypes.java]: +1, will change the name

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj (hunk 2 above)
Comment (nandorKollar) [reply lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj]: Make sense, will address this.

File: share/VERSION.txt (hunk 4 above)
Comment (nandorKollar) [reply share/VERSION.txt]: Resolving this comment, as the problem is addressed in #527

File: lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj (hunk 2 above)
Comment (nandorKollar) [reply lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj]: As of the other question: I was wondering on this question too, and I don't know. The author of the logical types seemed to find unimportant to add IDL representation for microsec, hence I didn't add it either for the new types. So far it seems that nobody else added this either, we can add the new token later if needed (to both types). Actually time type has one precision token too. The spec also has a duration type, which doesn't have any IDL token, though it seems that it isn't even implemented for Java, only for Ruby and C++.

File: doc/src/content/xdocs/spec.xml [hunk 5]
```diff
@@ -1530,6 +1530,26 @@ void initFPTable() {
         </p>
       </section>
 
+      <section>
+        <title>Local timestamp (millisecond precision)</title>
+        <p>
+          The <code>local-timestamp-millis</code> logical type represents a timestamp in a local timezone, regardless of what specific time zone is considered local, with a precision of one millisecond.
```
Comment (zivanfi) [inline doc/src/content/xdocs/spec.xml]: For local-timestamp-millis the semantics is specified but the interpretation of the numeric value is missing. For timestamp-millis, the interpretation of the numeric value that is specified, but an explanation of the semantics is missing. Could you please add both descriptions to both types? You can use the wording from here, if you would like to: https://github.com/apache/parquet-format/commit/fa6e468540d36122ca8a273e9c49425660c75360
